docs/types.rst

```diff
@@ -0,0 +1,145 @@
+variable (state and local) needs to be specified or at least known at
+compile-time. Viper provides several elementary types which can be combined
+to form complex types.
+
+In addition, types can interact with each other in expressions containing
+operators.
+
+Value Types
+===========
+
+The following types are also called value types because variables of these
+types will always be passed by value, i.e. they are always copied when they
+are used as function arguments or in assignments.
+
+.. index:: ! bool, ! true, ! false
+
+Booleans
+--------
+
+``bool``: The possible values are constants ``true`` and ``false``.
+
+Operators:
+
+*  ``not`` (logical negation)
+*  ``and`` (logical conjunction, "&&")
+*  ``or`` (logical disjunction, "||")
+*  ``==`` (equality)
+*  ``not ... == ... `` (inequality)
+*  ``!=`` (inequality)
+
+The operators ``or`` and ``and`` apply the common short-circuiting rules. This means that in the expression ``f(x) or g(y)``, if ``f(x)`` evaluates to ``true``, ``g(y)`` will not be evaluated even if it may have side-effects.
+
+.. index:: ! uint, ! int, ! integer
+
+
+Integers
+--------
+
+``num``:  a signed integer strictly between -2\*\*128 and 2\*\*128.
+
+Operators:
+
+* Comparisons: ``<=``, ``<``, ``==``, ``!=``, ``>=``, ``>`` (evaluate to ``bool``)
+* Arithmetic operators: ``+``, ``-``, unary ``-``, unary ``+``, ``*``, ``/``, ``%`` (remainder)
+
+
+Decimals
+--------
+``decimal``:  a decimal fixed point value with the integer component being a signed integer strictly between -2\*\*128 and 2\*\*128 and the fractional component being ten decimal places
+
+
+Time
+-----
+``timestamp``:  a timestamp value
+
+``timedelta``:  a number of seconds (note: two timedeltas can be added together, as can a timedelta and a timestamp, but not two timestamps)
+
+
+Value
+------
+``wei_value``:  an amount of wei
+
+``currency_value``:  an amount of currency
+
+
+
+.. _address:
+
+Address
+-------
+
+``address``: Holds a 20 byte value (size of an Ethereum address).
+
+
+.. _members-of-addresses:
+
+Members of Addresses
+^^^^^^^^^^^^^^^^^^^^
+
+* ``balance`` and ``send``
+
+It is possible to query the balance of an address using the property ``balance``
+and to send Ether (in units of wei) to an address using the ``send`` function:
+
+::
+
+    x: address
+
+    def foo(x: adress):
+        if (x.balance < 10 and self.balance >= 10):
+            x.send(10)
+
+
+.. index:: byte array, bytes32
+
+
+Fixed-size byte arrays
+----------------------
+
+``bytes32``: 32 bytes
+
+``type[length]``: finite list
+
+``bytes <= maxlen``: a byte array with the given maximum length
+
+
+Structs
+-------
+
+``{arg1:type, arg2:type...}``:  struct (can be accessed via struct.argname)
+
+
+.. index:: !mapping
+
+Mappings
+========
+
+Mapping types are declared as ``_ValueType[_KeyType]``.
+Here ``_KeyType`` can be almost any type except for mappings, a contract, or a struct.
+``_ValueType`` can actually be any type, including mappings.
```

Comment:
I thought nested mappings was disallowed e.g. `num[address][address]` doesn't work?

Reply:
As of now the following works:
```
hm: address[num][num]

def foo(x: num) -> address:
    self.hm[1][1] = self
    return self.hm[1][1]
```
Is that what you're talking about?

Reply:
That's a double-ly nested array, which I wasn't sure about it being explicitly allowed, but didn't hear that it was disallowed. I just tested:
```python
test: num[address][address]

def foo(x: address, y: address) -> num:
    return self.test[x][y]
```

No problems, which means I must be mistaken, but I could've sworn I remembered that @vbuterin said it should not be allowed because it would create an infinite keyspace or something. Anyways, I tested 10 levels deep and it still compiles:
```python
test: num[address][address][address][address][address][address][address][address][address][address]

def foo(x0: address, x1: address, x2: address, x3: address, x4: address, x5: address, x6: address, x7: address, x8: address, x9: address) -> num:
    return self.test[x0][x1][x2][x3][x4][x5][x6][x7][x8][x9]
```